### ISSUE TITLE
Einbetten von YouTube Videos im Blog

### ISSUE THREAD

**Issue by Rolf-B** [ckruse/cforum_ex#380]
Ich weiß nicht ob das in die Issues gehört.

Jedenfalls scheint das Einbetten nicht zu gehen wegen iframe Policy, Chrome sagt nacheinander:

```
Refused to frame 'https://www.youtube.com/' because it violates the following Content Security Policy directive: "frame-src wiki.selfhtml.org".
Refused to frame 'https://youtu.be/' because it violates the following Content Security Policy directive: "frame-src wiki.selfhtml.org".
Refused to frame 'https://www.youtube-nocookie.com/' because it violates the following Content Security Policy directive: "frame-src wiki.selfhtml.org".
```

Sicherlich ist das Absicht, aber ist wenn man in einem Blog ein Video einbetten möchte? Grundsätzlich verpönt?

**Comment by ckruse** [comment]
Das Einbetten von YouTube-Videos ist rechtlich grundsätzlich problematisch. Ich verstehe den Wunsch, jedoch muss hier mindestens die Heise-Zweiklick-Lösung benutzt werden.

That said: das löst natürlich nicht das Problem mit den CSP-Headern. Hm. Schwierig. Ich könnte die CSP-Header im Weblog deaktivieren, aber das gilt dann auch für die Kommentare. Oder ich könnte YouTube als erlaubte Quelle hinzufügen.

**Comment by Rolf-B** [comment]
Ok, was bedeutet das nun? Erstmal im Verein drüber reden? Ist die Heise-Lösung überhaupt etwas, was einbaubar ist oder schießt das total quer? Magst du zum nächsten Stammtisch kommen, dass wir drüber im größeren Kreis reden können?

Du sagst, das gilt dann auch für Kommentare. Grad bekomm ich Angst. Matthias sagte mir, dass HTML im Text etwas sei, das nur im Blog geht. Etwa  auch in den Kommentaren? **Das** klingt nicht gut; die Kommentare sollten nicht mehr können als normale Forenantworten. Denn Blogthreads zu erzeugen ist ein Sonderrecht für Leute besonderen Vertrauens. Drauf antworten kann aber jeder.

_Rolf_

**Comment by ckruse** [comment]
In Kommentaren ist nur erlaubt, was auch in Foren-Beiträgen erlaubt ist. Das bedeutet aber auch, dass User z.B. externe Bilder einbinden können, oder es ist möglich, dass ich eine Sicherheitslücke im Code habe. Die CSP-Header sind eine Absicherungs-Mechanik dagegen.

Bzgl der Heise-Zweiklick-Lösung: eine einfache Klick-to-Consent-Lösung ist sicherlich machbar. Nur für JS-Abschalter bliebe sie dann halt nur ein Link.

**Comment by Rolf-B** [comment]
Genau das meinte ich ja. Wenn wir iframes oder die YT Zweiklick Lösung für Blog-Beiträge erlauben, würden wir es zwangsweise auch in den Kommentare erlauben. Das "How To" kann sich dann jeder anschauen, der auf den Blogbeitrag mit Zitat antwortet. Und davor hätte ich Angst. So wie Du.

Wenn wir hier für Blogs mehr ermöglichen wollen - sprich: Im Blog geht mehr als im Forum - müsste man also zwischen "Was darf man im Erstposting" und "Was dürfen die Antworter*innen" unterscheiden. Das muss ja nicht spezifisch für das Blog-Forum programmiert werden. In meiner Naivität stelle ich mir zwei Attribute in der Konfiguration je Forum vor: RegularFormatter und FirstPostingFormatter. Der RegularFormatter greift als Default und der FirstPostingFormatter für das erste Posting eines Threads. Er wäre nur im Blog hinterlegt; oder generell: In Foren, wo die Thread-Eröffnung nur den Forenbetreibern möglich ist. In den anderen Foren wäre der FirstPostingFormatter NULL (und fällt im Code auf Regular zurück) oder es wird einfach der gleiche Wert wie für den RegularFormatter hinterlegt. Wie gesagt: das mag naiv und zu einfach gedacht sein. Aber es wäre meine erste Idee.

Was genau wir ermöglichen wollen, sprich: wo unterscheidet sich der FirstPostingFormatter, müssen wir noch diskutieren. Das mache ich Mittwoch zum Stammtischthema. Ein YT-Video per Zweiklick, ja, ich denke schon. Ein anderswo gehostetes MP4 Video? Anderswo gehostete Bilder? Im Wiki dürfte man Medien von Wikimedia Commons nutzen, meinte Matthias (weil das Wiki ein Wikimedia-Projekt ist); das dürfte im Forum anders sein. Was mach ich also, wenn ich, so wie Una Kravets in ihrem Blog, eine Animation als MP4 darstellen möchte? Kriegen wir sowas hin? Oder müsste ich die Animation als GIF oder animiertes SVG reproduzieren?

Ich habe keine Ahnung, welche Aufwände wir hier für Dich generieren würden.

**Comment by ckruse** [comment]
> Wenn wir iframes oder die YT Zweiklick Lösung für Blog-Beiträge erlauben, würden wir es zwangsweise auch in den Kommentare erlauben.

Nein, du hast mich falsch verstanden. In den KOMMENTAREN ist das erlaubt, was im Forum erlaubt ist. In den BLOG-ARTIKELN ist deutlich mehr erlaubt.

> Wenn wir hier für Blogs mehr ermöglichen wollen

Das ist bereits von Anfang an der Fall gewesen. In Blog-Beiträgen ist im Grunde jegliches HTML erlaubt.

Die YT-Zweiklick-Lösung ist deshalb ein Sonderfall, weil sie JS benötigt, das eingebunden werden muss – und da hakt es bereits wieder mit den CSP-Headern, wenn ich das nicht durch die Software ausliefere.

> Was mach ich also, wenn ich, so wie Una Kravets in ihrem Blog, eine Animation als MP4 darstellen möchte?

Hrm. Schwierig. Ich könnte einen Host wie z.B. `src.selfhtml.org` zu den erlaubten Hosts hinzufügen.

> Anderswo gehostete Bilder?

Davon halte ich gar nichts. Wir haben das in der Vergangenheit erlaubt, und das hat zu echt vielen kaputten Bildern sowohl im Forum als auch im Blog geführt. Ich hätte lieber alles in der Bilder-Sammlung vom CForum.

**Comment by Rolf-B** [comment]
src.selfhtml.org - okay?!
An der Stelle fehlt mir der Überblick über die Speicherstruktur des Selfraums und den diversen Wegen, irgendwo Dateien zu hinterlegen. Da gibt's mutmaßlich keine Doku zu, oder?

**Comment by ckruse** [comment]
Nein, nicht mehr. Ich hatte mal einen Artikel dazu geschrieben, aber der ist mit dem SELF-Aktuell-Raum beerdigt worden. Und es hat sich auch einiges geändert. Im Grunde hatte ich aber src.selfhtml.org bei seiner Einführung als SELF-CDN gedacht. Ist aber nach dem Server-Crash auch massiv umstrukturiert worden.

**Comment by Rolf-B** [comment]
Okay, dedlfix wird mir erklären wie der GIT-Zugang zu src.selfhtml geht.

Was ich noch fragen wollte, was aber nicht mehr wirklich zum Issue gehört: Ist die "Blog" Eigenschaft des /weblog Forums hardcoded oder ist das irgendwo pro Forum konfigurierbar? Hintergrund der Überlegung war: Könnte man eine "in Arbeit" Vorstufe des Blogs schaffen, wo man Beiträge zwischenspeichern kann ohne sie zu veröffentlichen? Mir fällt jetzt allerdings wieder auf, dass da eine wesentliche Sache fehlt: Ein Blog-Beitrag kann nur in der blog-Subdomain angelegt werden und hat ein anderes Eröffnungs-Form. Deswegen war die Überlegung wohl Unsinn, es sei denn, man erzeugt eine dev-blog Subdomain für Blogbeiträge die in Arbeit sind. Aber das scheint mir wenig sinnvoll.

Trotzdem, die Feature-Überlegung bleibt - als Workflow-Erleichterung. Kann man es hinbekommen, dass man Blog-Beiträge erzeugt ohne sie zu veröffentlichen? Denn im Moment muss ich, um die Formatierung zu testen, einen neuen Beitrag anlegen, den Inhalt hineinkopieren und gucken, ob's passt. Und dann muss ich den Arbeitsstand wieder herausnehmen und anderswo speichern, wenn ich noch nicht fertig bin.

**Comment by ckruse** [comment]
Um zunächst die Frage zu beantworten: die Blog-Eigenschaft des Subforums hängt an einem Attribut namens `type`. Damit aber mehrere Blog-Foren betrieben werden können, muss noch eine Möglichkeit geschaffen werden, Blog-Foren an Domains zu binden. Zur Zeit wird auf allen `blog.*`-Domains einfach das erste Blog-Forum genommen, dass gefunden wird. Das ist aber nicht sonderlich aufwendig.

Bzgl der unveröffentlichten Blog-Beiträge: da habe ich auch schon drüber nachgedacht. Das ist natürlich machbar, ich habs mir nur bisher gespart ;-)

**Comment by Rolf-B** [comment]
Mit einer "Draft" Option wären mehrfache Blogs wohl nachrangig.

**Comment by ckruse** [comment]
Es gibt jetzt einen Draft-Status für Nachrichten: <https://github.com/ckruse/cforum_ex/commit/75eb6c8bd215e87717993470bfa6662bb802a721>

Der Draft-Status kann nur für OPs im Weblog gesetzt werden. Er sorgt dafür, dass die Threads nicht in den Listen-Ansichten auftauchen, aber sie können direkt aufgerufen werden und sie tauchen in der Suche auf.

Das Feature ist noch nicht live, ich muss das erst noch testen. Jetzt ist aber erstmal Rasen mähen dran.

**Comment by Rolf-B** [comment]
Rasen mähen und 7 Uhr morgens? Wow ;)

Aber, super. Ich habe jetzt auch Git-Zugang zu src.selfhtml.org. Wenn Du diese URL als framesrc bzw. als Quelle für eingebundene Medien (das wäre die media-src CSP, scheint mir), wäre das Ergebnis für den Moment wohl perfekt. Die Zweiklick-Lösung für Youtube wäre irgendwann auch nicht schlecht, aber das werde ich erstmal mit einem einfachen Standbild lösen, das nach YT verlinkt.

**Comment by ckruse** [comment]
Ich möchte dringend davon abraten, große Binärdateien in ein Git-Repo einzuchecken. Das funktioniert zwar irgendwie, aber nicht besonders gut ;-)

**Comment by Rolf-B** [comment]
Die beiden MP4 sind 172kB und 262kB (kB, nicht KiB)
Was ist die Alternative? Bereitstellen eine FTP Zugangs? Die Übertragung per Git soll doch ein Tracking ermöglichen, wer wann was gemacht hat.

**Comment by ckruse** [comment]
Mir ist das egal. Ich muss das weder nutzen noch administrieren. Ich habe das nur zu bedenken gegeben.

**Comment by ckruse** [comment]
So. Ich habe das Draft-Feature online geschaltet (der Build läuft gerade, dauert noch ein paar Minuten) und den `media-src`-Header ergänzt.

**Comment by Rolf-B** [comment]
Ooookeeeeeh - ich habe meinen Beitrag als Draft gespeichert. Nun ist er weg. Wie rufe ich Drafts auf? Nochmal "Neuer Artikel" klicken hat's nicht gebracht.

**Comment by ckruse** [comment]
Du kannst ihn nachwievor über die URL aufrufen, und wenn du den Administrator-Modus einschaltest kannst du ihn auch noch sehen.

**Comment by Rolf-B** [comment]
Sorry, aber das ist zu kurz gedacht. 
- "Notieren" der URL für neue Blogposts geht nicht, die heißen ja "new". Wie komm ich als Autor an die URL?
- Blog-Autoren sind nicht zwingend Moderatoren. Entwürfe von User xyz müssen für User xyz auch ohne Admin-Sicht erreichbar sein.
- Wenn die Admin-Sicht den Blog beeinflusst, sollte man sie dort auch schalten können. Derzeit muss man ins Forum, die Admin-Sicht einschalten und in den Blog zurück, um sie zu haben. Das ist kontra-intuitiv. Aber, wie gesagt, für Entwürfe sollte man die Admin-Sicht nicht brauchen.
- Entwürfe müssen als solche dargestellt sein. Im derzeitigen Weblog-Forum ist das nicht der Fall, im Blog auch nicht. Wenn das zu kompliziert wird, ok, dann müssen wir das organisatorisch lösen und solche Posts mit "Entwurf" betiteln.

Noch was anderes seh ich gerade:
- Im Weblog-Forum muss ich ja "srt=yes" einschalten, um den Entwurf zu sehen. Gehe ich in diesem Modus auf die Blog-Seite, bleibt diese Option in der URL, und "Neuer Artikel" funktioniert nicht. "Neuer Artikei" erzeugt dann die URL `https://blog.selfhtml.org/?srt=yesnew?srt=yes`. Ich muss die URL editieren oder erst wieder ins Forum zurück, um gelesene Beiträge zu verstecken.
- Analoges gilt für den Admin-Modus. view_all=yes klemmt sich ebenfalls vorn in die URL und verhindert z.B. die Detailansicht eines Blogeintrags oder sein Bearbeiten. Man muss es für Entwürfe von Hand aus der URL entfernen.

**Comment by ckruse** [comment]
Die URL wird dir bekannt, wenn du den Thread speicherst. Mach dir ein Bookmark auf die URL, dann findest du den Thread auch wieder, wenn du keinen Zugriff auf die Admin-Ansicht hast.

**Comment by ckruse** [comment]
So. Die Draft-Threads werden jetzt mit „Entwurf:“ geprefixed. Die URL-Bugs sind auch gefixed. Die Admin-Ansicht wird jetzt auch im Weblog verlinkt; dass sie da nicht war, war eh ein Bug. Hab ich wohl schlicht vergessen.

Außerdem werden Drafts nicht mehr vor dem Owner versteckt.

Ist in ein paar Minuten online.

**Comment by Rolf-B** [comment]
Supi!

**Comment by Rolf-B** [comment]
Martin meinte, dass Entwürfe als neue Beiträge im Forum gezählt würden, die man dann aber nicht finden könne....

**Comment by ckruse** [comment]
Ja, da hatte er recht. Ist aber bereits seit 5.6.62 gefixed.
